### PR TITLE
feat(parser): support for CloudFormation Custom Resources

### LINKED ISSUE #2295: Feature request: Parser support for custom resource
### Use case

CFN Custom resource can trigger a lambda function with a set of events: create, delete and update.
There's a clear envelope and user payload which makes it a classic candidate for the parser.
I printed the actual events and wrote an event generator for testing purposes in my own application.
here's the gist: https://gist.github.com/ran-isenberg/9a49fad3d607245f7d615696353d1a36

### Solution/User Experience
```
@event_parser(model=UserModel, envelope=envelopes.CustomResourceEnvelope)
def handler(event: UserModel, context: LambdaContext):
    assert event.password1 == event.password2
```

### Alternative solutions

_No response_

### Acknowledgment

- [X] This feature request meets [Lambda Powertools Tenets](https://awslabs.github.io/aws-lambda-powertools-python/latest/#tenets)
- [X] Should this be considered in other Lambda Powertools languages? i.e. [Java](https://github.com/awslabs/aws-lambda-powertools-java/), [TypeScript](https://github.com/awslabs/aws-lambda-powertools-typescript/), and [.NET](https://github.com/awslabs/aws-lambda-powertools-dotnet/)

### PR DESCRIPTION
**Pull request by ran-isenberg**
**Issue number:** #2295 

### Changes

* [x] added 3  parser models: create/update/delete custom resource event source, 
* [x] added 3 json test files,
* [x] added tests and added documentation of models.

### User experience

```python
from aws_lambda_powertools.utilities.parser import event_parser
from aws_lambda_powertools.utilities.parser.models import (
    CloudFormationCustomResourceCreateModel,
    CloudFormationCustomResourceDeleteModel,
    CloudFormationCustomResourceUpdateModel,
)
from aws_lambda_powertools.utilities.typing import LambdaContext


@event_parser(model=CloudFormationCustomResourceCreateModel)
def handle_create_custom_resource(event: CustomResourceCreateModel, _: LambdaContext):
    assert event.request_type == "Create"

@event_parser(model=CloudFormationCustomResourceUpdateModel)
def handle_update_custom_resource(event: CustomResourceUpdateModel, _: LambdaContext):
    assert event.request_type == "Update"

@event_parser(model=CloudFormationCustomResourceDeleteModel)
def handle_delete_custom_resource(event: CustomResourceDeleteModel, _: LambdaContext):
    assert event.request_type == "Delete"


```

Using extended model:

```python
class MyModel(BaseModel):
    MyProps: str


class MyCustomResource(CloudFormationCustomResourceCreateModel):
    resource_properties: MyModel = Field(..., alias="ResourceProperties")


@event_parser(model=MyCustomResource)
def handle_create_custom_resource_extended_model(event: MyCustomResource, _: LambdaContext):
    assert event.request_type == "Create"
    assert event.resource_properties.MyProps == "ss"

```

## Checklist

If your change doesn't seem to apply, please leave them unchecked.

* [x] [Meet tenets criteria](https://awslabs.github.io/aws-lambda-powertools-python/#tenets)
* [x] I have performed a self-review of this change
* [x] Changes have been tested
* [x] Changes are documented
* [x] PR title follows [conventional commit semantics](https://github.com/awslabs/aws-lambda-powertools-python/blob/develop/.github/semantic.yml)

<details>
<summary>Is this a breaking change?</summary>
No.
**RFC issue number**:

Checklist:

* [ ] Migration process documented
* [ ] Implement warnings (if it can live side by side)

</details>

## Acknowledgment

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

**Disclaimer**: We value your time and bandwidth. As such, any pull requests created on non-triaged issues might not be successful.

### CODE REVIEW

**Comment by ran-isenberg** [comment]
@heitorlessa  @leandrodamascena  pr is ready. let me know what you think #2335

**Comment by ran-isenberg** [comment]
@heitorlessa  @leandrodamascena  fixed the code for older python support

**Comment by leandrodamascena** [comment]
Hi @ran-isenberg! Thanks for sending the PR and sorry for the late reply, I was attending a meetup in another country.

I'll look at it tomorrow; if all goes well, we can include it in the next release.

**Comment by heitorlessa** [comment]
Assigning to Leandro as he shared he's taking it @leandrodamascena

File: aws_lambda_powertools/utilities/parser/models/cloudformation_custom_resource.py
Comment (leandrodamascena) [inline aws_lambda_powertools/utilities/parser/models/cloudformation_custom_resource.py]: Hi @ran-isenberg! According to the documentation, this field is not required, so it should be Optional, right?
https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/crpg-ref-requests.html

>ResourceProperties
This field contains the contents of the Properties object sent by the template developer. Its contents are defined by the custom resource provider.
Required: No
Type: JSON object

Comment (leandrodamascena) [inline aws_lambda_powertools/utilities/parser/models/cloudformation_custom_resource.py]:
```suggestion
    resource_properties: Dict[str, Any] | Type[BaseModel] | None = Field(None, alias="ResourceProperties")
```

File: aws_lambda_powertools/utilities/parser/models/cloudformation_custom_resource.py
Comment (leandrodamascena) [inline aws_lambda_powertools/utilities/parser/models/cloudformation_custom_resource.py]:
```suggestion
    old_resource_properties: Dict[str, Any] | Type[BaseModel] = Field(..., alias="OldResourceProperties")
```

File: docs/utilities/parser.md
Comment (leandrodamascena) [inline docs/utilities/parser.md]: Please consider to move this to the line before `CloudwatchLogsModel` so we can maintain alphabetical order.

File: aws_lambda_powertools/utilities/parser/models/cloudformation_custom_resource.py
Comment (leandrodamascena) [inline aws_lambda_powertools/utilities/parser/models/cloudformation_custom_resource.py]:
```suggestion
from __future__ import annotations

from typing import Any, Dict, Type

```

**Review by leandrodamascena — CHANGES_REQUESTED** [review]
Hi @ran-isenberg! We need to make some updates before we merge this.

**Review by leandrodamascena — DISMISSED** [review]
Hi @ran-isenberg! We need to make some updates before we merge this.

File: aws_lambda_powertools/utilities/parser/models/cloudformation_custom_resource.py
Comment (ran-isenberg) [reply aws_lambda_powertools/utilities/parser/models/cloudformation_custom_resource.py]: yeah i'l fix it.

File: docs/utilities/parser.md
Comment (ran-isenberg) [reply docs/utilities/parser.md]: np

**Comment by ran-isenberg** [comment]
@leandrodamascena  I've pushed your suggestions

File: aws_lambda_powertools/utilities/parser/models/cloudformation_custom_resource.py
Comment (ran-isenberg) [reply aws_lambda_powertools/utilities/parser/models/cloudformation_custom_resource.py]: fixed

**Comment by heitorlessa** [comment]
hey @ran-isenberg could you update the PR body with the new name so we don't forget?

Having a look at the rest now

**Comment by ran-isenberg** [comment]
> hey @ran-isenberg could you update the PR body with the new name so we don't forget?
> 
> Having a look at the rest now

It seems older python code doesnt like the '|' instead of Union. what do you suggest?
I've updated the PR body

**Comment by heitorlessa** [comment]
You can import from future: `from __future__ import annotations`. That should work. Pydantic has support to handle postpone evaluation of annotations.

https://peps.python.org/pep-0563/

**Comment by ran-isenberg** [comment]
> You can import from future: `from __future__ import annotations`. That should work. Pydantic has support to handle postpone evaluation of annotations.
> 
> https://peps.python.org/pep-0563/

Yes, I added that import, still no go. In other models we use Optional and Union which I think are more readable than the pipes '|'.
Not sure how to continue here.

**Comment by heitorlessa** [comment]
let me have a look ;) just finishing my last meeting of the day \o/

**Comment by heitorlessa** [comment]
I see what's happening, you've just hit a bug in the grammar, and they fixed it in 3.10 only :/ https://bugs.python.org/issue42233

I'll revert to Union and make any necessary changes to get this merged.

**Comment by ran-isenberg** [comment]
> I see what's happening, you've just hit a bug in the grammar, and they fixed it in 3.10 only :/ https://bugs.python.org/issue42233
> 
> I'll revert to Union and make any necessary changes to get this merged.

oh wow, that's a nasty one.

**Comment by heitorlessa** [comment]
Ready to merge - these are the changes I've made. Waiting for CI to complete then merge.

---

- [x] Rename test events to prefix CloudFormation (we wouldn't know otherwise in the future)
- [x] Revert union type as `|` to Union due to [bug not backported](https://bugs.python.org/issue42233)
- [x] Docs: improved Model description
  - **Before**: Lambda Event Source payload for AWS CloudFormation 'create' custom resource
  - **After**: Lambda Event Source payload for AWS CloudFormation `CREATE` operation
- [x] Rename test name to match model name (cloudformation_custom_resource)
- [x] Fix Model type in `resource_properties` to allow model override (`Type[BaseModel]` -> `BaseModel`). `Type[BaseModel]` would be for envelopes only
- [x] Refactor tests as unit tests - keeps tests focused on Model validation, less boilerplate, and allows us to update test data in the future
  - Get rid of `event_parser`, `handle_` functions, and keep contained within the Model we're testing
  - Rename test to match our standard
  - Replace hard coded value with pointer to allow value changing in the future

**Review by heitorlessa — APPROVED** [review]

**Comment by heitorlessa** [comment]
Thanks a lot @ran-isenberg for the contribution, much appreciated! The major change I've made that's worth pointing out is to refactor those tests as unit tests - here's the difference in a nutshell.


**BEFORE**

```python
@event_parser(model=CloudFormationCustomResourceDeleteModel)
def handle_delete_custom_resource(event: CloudFormationCustomResourceDeleteModel, _: LambdaContext):
    assert event.request_type == "Delete"
    assert event.request_id == "xxxxx-d2a0-4dfb-ab1f-xxxxxx"
    assert event.service_token == "arn:aws:lambda:us-east-1:xxx:function:xxxx-CrbuiltinfunctionidProvi-2vKAalSppmKe"
    assert (
        str(event.response_url)
        == "https://cloudformation-custom-resource-response-useast1.s3.amazonaws.com/7F%7Cb1f50fdfc25f3b"
    )
    assert event.stack_id == "arn:aws:cloudformation:us-east-1:xxxx:stack/xxxx/271845b0-f2e8-11ed-90ac-0eeb25b8ae21"
    assert event.logical_resource_id == "xxxxxxxxx"
    assert event.resource_type == "Custom::MyType"
    assert event.resource_properties == {
        "ServiceToken": "arn:aws:lambda:us-east-1:xxxxx:function:xxxxx",
        "MyProps": "ss",
    }

def test_delete_trigger_event():
    event_dict = load_event("cloudformationCustomResourceDelete.json")
    handle_delete_custom_resource(event_dict, LambdaContext())
```

**AFTER**


```python
def test_cloudformation_custom_resource_delete_event():
    raw_event = load_event("cloudformationCustomResourceDelete.json")
    model = CloudFormationCustomResourceDeleteModel(**raw_event)

    assert model.request_type == raw_event["RequestType"]
    assert model.request_id == raw_event["RequestId"]
    assert model.service_token == raw_event["ServiceToken"]
    assert str(model.response_url) == raw_event["ResponseURL"]
    assert model.stack_id == raw_event["StackId"]
    assert model.logical_resource_id == raw_event["LogicalResourceId"]
    assert model.resource_type == raw_event["ResourceType"]
    assert model.resource_properties == raw_event["ResourceProperties"]
```

This now focuses on the Model validation per se, no hard code data allowing us to have a schedule routine to generate real event to spot regressions in the future, and reduce boilerplate.

If we use the same idea and try to test the Create test with a Custom model, we now only have to test the model per se:

```python
def test_cloudformation_custom_resource_create_event_custom_model():
    class MyModel(BaseModel):
        MyProps: str

    class MyCustomResource(CloudFormationCustomResourceCreateModel):
        resource_properties: MyModel = Field(..., alias="ResourceProperties")

    raw_event = load_event("cloudformationCustomResourceCreate.json")
    model = MyCustomResource(**raw_event)

    assert model.resource_properties.MyProps == raw_event["ResourceProperties"].get("MyProps")
```

**Comment by ran-isenberg** [comment]
@heitorlessa nice one, makes more sense as a unit test

**Pull request merged**
